File: frontend/src/api/query.ts
import { client } from './client'
import type { QueryHistoryPage, QueryResult } from '../types'

export async function askQuery(
  nl_query: string,
  datasource_id: string | null,
  previous_query_log_id: string | null = null,
): Promise<QueryResult> {
  const { data } = await client.post<QueryResult>('/query/ask', {
    nl_query,
    datasource_id,
    previous_query_log_id,
  })
  return data
}

export async function runSql(
  sql: string,
  datasource_id: string | null,
  label: string | null = null,
): Promise<QueryResult> {
  const { data } = await client.post<QueryResult>('/query/run', {
    sql,
    datasource_id,
    label,
  })
  return data
}

export async function getHistory(page = 1, limit = 20): Promise<QueryHistoryPage> {
  const { data } = await client.get<QueryHistoryPage>('/query/history', {
    params: { page, limit },
  })
  return data
}

export async function getQuery(id: string): Promise<QueryResult> {
  const { data } = await client.get<QueryResult>(`/query/${id}`)
  return data
}

export async function deleteQuery(id: string): Promise<void> {
  await client.delete(`/query/${id}`)
}

File: backend/app/schemas/query.py
"""Query request/response schemas."""
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field

from app.ai.types import ChartConfig


class ColumnInfo(BaseModel):
    name: str
    type: str = "unknown"


class QueryRequest(BaseModel):
    nl_query: str = Field(min_length=1, max_length=2000)
    datasource_id: str | None = None
    # Previous turn for multi-turn "chat with your data" follow-ups.
    previous_query_log_id: str | None = None


class QueryResult(BaseModel):
    sql: str
    # "sql" for database sources, "dax" for Power BI sources (UI labels it).
    query_language: str = "sql"
    data: list[dict[str, Any]]
    columns: list[ColumnInfo]
    chart_config: ChartConfig
    insight: str = ""
    execution_time_ms: int = 0
    query_log_id: str | None = None
    from_cache: bool = False


class QueryHistoryItem(BaseModel):
    id: str
    natural_language: str
    generated_sql: str
    chart_type: str
    execution_time_ms: int
    created_at: str


class QueryHistoryPage(BaseModel):
    items: list[QueryHistoryItem]
    page: int
    limit: int
    total: int
